test/jdk/java/net/httpclient/HttpResponseLimitingTest.java

```diff
@@ -0,0 +1,275 @@
+                },
+                "the maximum number of bytes that are allowed to be consumed is exceeded");
+    }
+
+    static long[] insufficientCapacities() {
+        return new long[]{0, HttpServer.RESPONSE_BODY.length - 1};
+    }
+
+    @Test
+    void testSubscriberForCompleteConsumption() {
+
+        // Create the subscriber (with sufficient capacity)
+        ObserverSubscriber downstreamSubscriber = new ObserverSubscriber();
+        int sufficientCapacity = HttpServer.RESPONSE_BODY.length;
+        BodySubscriber<String> subscriber = BodySubscribers.limiting(downstreamSubscriber, sufficientCapacity);
+
+        // Emit values
+        subscriber.onSubscribe(DummySubscription.INSTANCE);
+        byte[] responseBodyPart1 = {HttpServer.RESPONSE_BODY[0]};
+        byte[] responseBodyPart2 = copyOfRange(HttpServer.RESPONSE_BODY, 1, HttpServer.RESPONSE_BODY.length);
+        List<ByteBuffer> buffers = toByteBuffers(responseBodyPart1, responseBodyPart2);
+        subscriber.onNext(buffers);
+
+        // Verify the downstream propagation
+        assertSame(buffers, downstreamSubscriber.lastBuffers);
+        assertNull(downstreamSubscriber.lastThrowable);
+        assertFalse(downstreamSubscriber.completed);
+
+    }
+
+    @Test
+    void testSubscriberForFailureOnExcess() {
+
+        // Create the subscriber (with insufficient capacity)
+        ObserverSubscriber downstreamSubscriber = new ObserverSubscriber();
+        int insufficientCapacity = 2;
+        BodySubscriber<String> subscriber = BodySubscribers.limiting(downstreamSubscriber, insufficientCapacity);
+
+        // Emit values
+        subscriber.onSubscribe(DummySubscription.INSTANCE);
+        byte[] responseBodyPart1 = {HttpServer.RESPONSE_BODY[0]};
+        byte[] responseBodyPart2 = copyOfRange(HttpServer.RESPONSE_BODY, 1, HttpServer.RESPONSE_BODY.length);
+        List<ByteBuffer> buffers = toByteBuffers(responseBodyPart1, responseBodyPart2);
+        subscriber.onNext(buffers);
+
+        // Verify the downstream propagation
+        assertNull(downstreamSubscriber.lastBuffers);
+        assertNotNull(downstreamSubscriber.lastThrowable);
+        assertEquals(
+                "the maximum number of bytes that are allowed to be consumed is exceeded",
+                downstreamSubscriber.lastThrowable.getMessage());
+        assertFalse(downstreamSubscriber.completed);
+
+    }
+
+    private static List<ByteBuffer> toByteBuffers(byte[]... buffers) {
+        return Arrays.stream(buffers).map(ByteBuffer::wrap).collect(Collectors.toList());
+    }
+
+    /**
+     * An HTTP server always returning a fixed response containing a non-empty body.
+     */
+    private static final class HttpServer implements Runnable, AutoCloseable {
+
+        private static final byte[] RESPONSE_BODY = "random non-empty body".getBytes(CHARSET);
+
+        private static final byte[] RESPONSE = (
+                "HTTP/1.2 200 OK\r\n" +
```

Comment:
Did you mean `HTTP/1.1`?

Reply:
Fixed in 23f79a20476f784058801454a28d37f37df92982, replaced with `HttpServerAdapters` in 64190c5765dc486eafd1d37b9312b1256943ed8d.